### PR TITLE
Implement at::has_internal_overlap helper function

### PR DESCRIPTION
**Pull request by zou3519**
Stack:
* **#17926 Implement at::has_internal_overlap helper function**
* #17927 Error out on in-place (unary) ops on tensors that have internal overlap
* #17993 [easy] Delete dead code in THTensorMoreMath.cpp

On the way to #17935.

Checks if a tensor's sizes/strides indicate that multiple elements share
the same memory location. This problem in general is hard so
at::has_internal_overlap implements two heuristics and avoids solving
the general problem:

if a tensor is contiguous, it cannot have internal overlap
if a tensor has any zero strides, it does have internal overlap
otherwise, return MemOverlap::kTooHard to indicate that there might be
overlap, but we don't know.

Test Plan:
- I've exposed a _debug_has_internal_overlap that I test via
test_torch.py

Future:
- Working on disabling partial aliasing because that is the easiest
thing to do for now: we should error out on in-place ops on tensors
that have internal overlap.
- There are some other heuristics we can apply [like
this](https://github.com/pytorch/pytorch/blob/86fbf17ba64f985a7d632576091a08f8344f7a24/aten/src/ATen/cuda/detail/IndexUtils.cu#L37-L71)
and
[this](https://github.com/pytorch/pytorch/blob/44fb22f9fe7c8eb315c715c0b36213f282a6835e/tools/autograd/templates/Functions.cpp#L1139-L1363)
but those are only useful when enabling partial aliasing so I haven't
merged those in yet.

gh-metadata: pytorch pytorch 17926 gh/zou3519/22/head

Differential Revision: [D14438858](https://our.internmc.facebook.com/intern/diff/D14438858)

### CODE REVIEW

**Comment by SsnL** [comment]
out of curiosity, what do you mean by "partial aliasing"?

**Comment by zou3519** [comment]
"partial aliasing" = tensor(s) have memory overlap. I should be consistent and start using "memory overlap" because the term "partial aliasing" is confusing, as you've pointed out.

**Comment by SsnL** [comment]
Thanks for the clarification. Unfortunately, I am still not fully understanding why the two referenced heuristics can't be used here. From my understanding, they detect whether a tensor absolutely does not have internal memory overlap, which seems useful to me in cases of returning `MemOverlap::kNo`.

**Comment by zou3519** [comment]
Right now, my plan is to detect if a tensor absolutely does have internal overlap. If so, then we error out.

There are three main cases:
- tensor absolutely does have internal overlap
- tensor absolutely doesn't have internal overlap
- tensor may have internal overlap.

If a tensor "may have internal overlap" or "absolutely doesn't have internal overlap", then the plan is to not do anything right now. Because of this, it isn't useful to differentiate between "may have internal overlap" and "absolutely doesn't have internal overlap" (that is what the two helper functions do).

We might want to throw a warning in the "may have internal overlap" case -- if we choose to do this those helper functions will be useful, but I haven't thought very hard about this case and I am worried we might spam people with warnings.

**Comment by SsnL** [comment]
@zou3519 Thanks! This makes perfect sense!  It seems that we need the numpy thing eventually here.

**Comment by zou3519** [comment]
I'll go write up what I wrote there into an issue (because all we have appear to be bug reports and I realized it is ambiguous what our plan is right now). The numpy thing has a runtime that is exponential in the number of dimensions (yikes!) and it doesn't cover all cases, so I think we want to implement a few heuristics (like the "Detecting Memory Overlap Within A Strided Tensor") and ignore solving the general case

File: aten/src/ATen/MemoryOverlap.cpp [hunk 1]
```diff
@@ -0,0 +1,34 @@
+#include <ATen/MemoryOverlap.h>
+#include <c10/core/Layout.h>
+#include <c10/core/Backend.h>
+
+namespace at {
+
+MemOverlap has_internal_overlap(const Tensor& tensor) {
+  auto* t = tensor.unsafeGetTensorImpl();
+
+  AT_ASSERT(layout_from_backend(tensorTypeIdToBackend(t->type_id())) == kStrided);
```
Comment (ezyang) [inline aten/src/ATen/MemoryOverlap.cpp]: Err, is there really no way to get the layout directly from a tensor?

File: aten/src/ATen/MemoryOverlap.h [hunk 2]
```diff
@@ -0,0 +1,13 @@
+#pragma once
+
+#include <ATen/ATen.h>
+
+namespace at {
+
+enum MemOverlap { kNo, kYes, kTooHard };
```
Comment (ezyang) [inline aten/src/ATen/MemoryOverlap.h]: Make this an enum class. Also, the prevailing naming convention for enum members is all-caps, not k-prefixed (which is for non-enum constants. For example kCPU = DeviceType::CPU)

Comment (ezyang) [reply aten/src/ATen/MemoryOverlap.h]: It's pretty clear to me what this says, but for others, could we add a comment for this? Also mention that you need to update the Python test if you renumber this.

**Review by ezyang — APPROVED** [review]
Accepted with comments

File: aten/src/ATen/MemoryOverlap.cpp (hunk 1 above)
Comment (zou3519) [reply aten/src/ATen/MemoryOverlap.cpp]: Oh, there's a tensor.layout() function. I should use it

**Pull request closed without merging**